File: app/src/main/java/com/nfcalarmclock/NacDayOfWeek.java
package com.nfcalarmclock;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.LinearLayout.LayoutParams;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

/**
 * A button that consists of an image to the left, and text to the right of it.
 */
public class NacDayOfWeek
	implements NacDayButton.OnDayChangedListener
{

	/**
	 * Listen for when the day of week is changed.
	 *
	 * Returning True means that the listener handled the event, and False means
	 * that the NacDayOfWeek class will handle the click event, and execute the
	 * default action.
	 */
	public interface OnWeekChangedListener
	{
		public boolean onWeekChanged(NacDayButton button, NacCalendar.Day day);
	}

	/**
	 * Day of week view.
	 */
	private LinearLayout mDayOfWeekView;

	/**
	 * Day of week changed listener.
	 */
	private NacDayOfWeek.OnWeekChangedListener mOnWeekChangedListener;

	/**
	 */
	public NacDayOfWeek(LinearLayout view)
	{
		this.mDayOfWeekView = view;
		this.mOnWeekChangedListener = null;
		int count = view.getChildCount();

		for (int i=0; i < count; i++)
		{
			NacDayButton button = (NacDayButton) view.getChildAt(i);
			button.setOnDayChangedListener(this);
		}
	}

	/**
	 * @return True if any days are selected, and False otherwise.
	 */
	public boolean areDaysSelected()
	{
		return !this.getDays().isEmpty();
	}

	/**
	 * Convert a particular day to its corresponding view ID.
	 */
	private int dayToId(NacCalendar.Day day)
	{
		switch (day)
		{
			case SUNDAY:
				return R.id.dow_sun;
			case MONDAY:
				return R.id.dow_mon;
			case TUESDAY:
				return R.id.dow_tue;
			case WEDNESDAY:
				return R.id.dow_wed;
			case THURSDAY:
				return R.id.dow_thu;
			case FRIDAY:
				return R.id.dow_fri;
			case SATURDAY:
				return R.id.dow_sat;
			default:
				break;
		}
		return 0;
	}

	/**
	 * @return The day button given a particular day.
	 */
	public NacDayButton getDayButton(NacCalendar.Day day)
	{
		LinearLayout view = this.getDayOfWeekView();
		int id = this.dayToId(day);
		return (NacDayButton) view.findViewById(id);
	}

	/**
	 * @return A list of all the day buttons.
	 */
	public List<NacDayButton> getDayButtons()
	{
		LinearLayout view = this.getDayOfWeekView();
		int count = view.getChildCount();
		List<NacDayButton> buttons = new ArrayList<>();

		for (int i=0; i < count; i++)
		{
			NacDayButton b = (NacDayButton) view.getChildAt(i);
			buttons.add(b);
		}

		return buttons;
	}

	/**
	 * @return The day of week view.
	 */
	public LinearLayout getDayOfWeekView()
	{
		return this.mDayOfWeekView;
	}

	/**
	 * @return The alarm days.
	 */
	public EnumSet<NacCalendar.Day> getDays()
	{
		EnumSet<NacCalendar.Day> days = EnumSet.noneOf(NacCalendar.Day.class);

		for (NacCalendar.Day d : NacCalendar.WEEK)
		{
			if (this.isDayEnabled(d))
			{
				days.add(d);
			}
		}

		return days;
	}

	/**
	 * @return The day of week on click listener.
	 */
	public NacDayOfWeek.OnWeekChangedListener getOnWeekChangedListener()
	{
		return this.mOnWeekChangedListener;
	}

	/**
	 * Convert an view ID to its corresponding day.
	 */
	private NacCalendar.Day idToDay(int id)
	{
		switch (id)
		{
			case R.id.dow_sun:
				return NacCalendar.Day.SUNDAY;
			case R.id.dow_mon:
				return NacCalendar.Day.MONDAY;
			case R.id.dow_tue:
				return NacCalendar.Day.TUESDAY;
			case R.id.dow_wed:
				return NacCalendar.Day.WEDNESDAY;
			case R.id.dow_thu:
				return NacCalendar.Day.THURSDAY;
			case R.id.dow_fri:
				return NacCalendar.Day.FRIDAY;
			case R.id.dow_sat:
				return NacCalendar.Day.SATURDAY;
			default:
				break;
		}
		return null;
	}

	/**
	 * @return True if the button is enabled and false if it is not.
	 */
	public boolean isDayEnabled(NacCalendar.Day day)
	{
		NacDayButton button = this.getDayButton(day);
		return button.isEnabled();
	}

	/**
	 */
	@Override
	public void onDayChanged(NacDayButton button)
	{
		int id = button.getId();
		NacCalendar.Day day = this.idToDay(id);
		NacDayOfWeek.OnWeekChangedListener listener =
			this.getOnWeekChangedListener();

		if (listener != null)
		{
			listener.onWeekChanged(button, day);
		}
	}

	/**
	 * Set the days that will be enabled/disabled.
	 *
	 * TODO This is doing more animating than necessary.
	 *      Only enable/disable if needs to be done.
	 *
	 * @param  days  The button days that will be enabled.
	 */
	public void setDays(EnumSet<NacCalendar.Day> days)
	{
		for (NacCalendar.Day d : NacCalendar.WEEK)
		{
			NacDayButton button = this.getDayButton(d);
			button.setOnDayChangedListener(null);

			if (days.contains(d))
			{
				button.enable();
			}
			else
			{
				button.disable();
			}

			button.setOnDayChangedListener(this);
		}
	}

	/**
	 * @see setDays
	 */
	public void setDays(int value)
	{
		EnumSet<NacCalendar.Day> days = NacCalendar.Days.valueToDays(value);
		this.setDays(days);
	}

	/**
	 * Set the day to start week on.
	 */
	public void setStartWeekOn(int start)
	{
		LinearLayout view = this.getDayOfWeekView();
		NacDayButton sunday = this.getDayButton(NacCalendar.Day.SUNDAY);
		NacDayButton monday = this.getDayButton(NacCalendar.Day.MONDAY);
		View firstChild = view.getChildAt(0);
		View lastChild = view.getChildAt(6);

		if (start == 1)
		{
			if (firstChild.getId() != monday.getId())
			{
				view.removeView(firstChild);
				view.addView(firstChild, 6);
			}
		}
		else
		{
			if (firstChild.getId() != sunday.getId())
			{
				view.removeView(lastChild);
				view.addView(lastChild, 0);
			}
		}
	}

	/**
	 * Set the listener for the week is changed.
	 */
	public void setOnWeekChangedListener(
		NacDayOfWeek.OnWeekChangedListener listener)
	{
		this.mOnWeekChangedListener = listener;
	}

}

File: app/src/main/java/com/nfcalarmclock/view/dayofweek/NacDayOfWeek.kt
package com.nfcalarmclock.view.dayofweek

import android.widget.LinearLayout
import com.nfcalarmclock.R
import com.nfcalarmclock.util.NacCalendar
import com.nfcalarmclock.util.NacCalendar.Day
import com.nfcalarmclock.view.dayofweek.NacDayButton.OnDayChangedListener
import java.util.EnumSet

/**
 * A button that consists of an image to the left, and text to the right of it.
 */
class NacDayOfWeek(

	/**
	 * Day of week view.
	 */
	private val dayOfWeekView: LinearLayout

) : OnDayChangedListener
{

	/**
	 * Listen for when the day of week is changed.
	 *
	 * Returning True means that the listener handled the event, and False means
	 * that the NacDayOfWeek class will handle the click event, and execute the
	 * default action.
	 */
	fun interface OnWeekChangedListener
	{
		fun onWeekChanged(button: NacDayButton, day: Day)
	}

	/**
	 * A list of all the day buttons.
	 */
	val dayButtons: List<NacDayButton>
		get()
		{
			// List of buttons
			val buttons: MutableList<NacDayButton> = ArrayList()

			// Count the number of views in the container
			val count = dayOfWeekView.childCount

			// Iterate over each view
			for (i in 0 until count)
			{
				// Get the button
				val b = dayOfWeekView.getChildAt(i) as NacDayButton

				// Add it to the list
				buttons.add(b)
			}

			// Return the list of buttons
			return buttons
		}

	/**
	 * The alarm days.
	 */
	val days: EnumSet<Day>
		get()
		{
			val days = Day.NONE

			// Iterate over each day in the week
			for (d in Day.WEEK)
			{
				// Get the button
				val button = getDayButton(d)

				// Check if the day is enabled
				if (button.button!!.isChecked)
				{
					// Add the day
					days.add(d)
				}
			}

			// Return the set of days
			return days
		}

	/**
	 * Day of week changed listener.
	 */
	var onWeekChangedListener: OnWeekChangedListener? = null

	/**
	 * Constructor
	 */
	init
	{
		// Setup buttons
		setupDayButtons()
	}

	/**
	 * Convert a particular day to its corresponding view ID.
	 */
	private fun dayToId(day: Day): Int
	{
		return when (day)
		{
			Day.SUNDAY -> R.id.dow_sun
			Day.MONDAY -> R.id.dow_mon
			Day.TUESDAY -> R.id.dow_tue
			Day.WEDNESDAY -> R.id.dow_wed
			Day.THURSDAY -> R.id.dow_thu
			Day.FRIDAY -> R.id.dow_fri
			Day.SATURDAY -> R.id.dow_sat
		}
	}

	/**
	 * Get the day button given a particular day.
	 *
	 * @return The day button given a particular day.
	 */
	private fun getDayButton(day: Day): NacDayButton
	{
		// Get the ID
		val id = dayToId(day)

		// Find the view
		return dayOfWeekView.findViewById(id)
	}

	/**
	 * Convert an view ID to its corresponding day.
	 */
	private fun idToDay(id: Int): Day?
	{
		return when (id)
		{
			R.id.dow_sun -> Day.SUNDAY
			R.id.dow_mon -> Day.MONDAY
			R.id.dow_tue -> Day.TUESDAY
			R.id.dow_wed -> Day.WEDNESDAY
			R.id.dow_thu -> Day.THURSDAY
			R.id.dow_fri -> Day.FRIDAY
			R.id.dow_sat -> Day.SATURDAY
			else -> null
		}
	}

	/**
	 * Called when a day is changed
	 */
	override fun onDayChanged(button: NacDayButton)
	{
		// Get the day or return if it is null
		val day = idToDay(button.id) ?: return

		// Call the listener
		onWeekChangedListener?.onWeekChanged(button, day)
	}

	/**
	 * Set the days that will be enabled/disabled.
	 *
	 * TODO This is doing more animating than necessary.
	 * Only enable/disable if needs to be done.
	 *
	 * @param  days  The button days that will be enabled.
	 */
	fun setDays(days: EnumSet<Day>)
	{
		// Iterate over each day in the week
		for (d in Day.WEEK)
		{
			// Get the button
			val button = getDayButton(d)

			// Clear the listener
			button.onDayChangedListener = null

			// Change the button's state depending on the days
			if (days.contains(d))
			{
				button.enable()
			}
			else
			{
				button.disable()
			}

			// Set the listener once again
			button.onDayChangedListener = this
		}
	}

	/**
	 * @see .setDays
	 */
	fun setDays(value: Int)
	{
		// Determine the days
		val days = Day.valueToDays(value)

		// Set the days
		this.setDays(days)
	}

	/**
	 * Set the day to start week on.
	 *
	 * TODO: Add more days
	 */
	fun setStartWeekOn(start: Int)
	{
		// Get the buttons that can be started on
		val sunday = getDayButton(Day.SUNDAY)
		val monday = getDayButton(Day.MONDAY)

		// Get the first and last children
		val firstChild = dayOfWeekView.getChildAt(0)
		val lastChild = dayOfWeekView.getChildAt(6)

		// Check which days will be started on
		if (start == 1)
		{
			if (firstChild.id != monday.id)
			{
				dayOfWeekView.removeView(firstChild)
				dayOfWeekView.addView(firstChild, 6)
			}
		}
		else
		{
			if (firstChild.id != sunday.id)
			{
				dayOfWeekView.removeView(lastChild)
				dayOfWeekView.addView(lastChild, 0)
			}
		}
	}

	/**
	 * Setup the day buttons.
	 */
	private fun setupDayButtons()
	{
		val context = dayOfWeekView.context

		// Get the days of week
		val week = context.resources.getStringArray(R.array.days_of_week)

		// Count the number of children (should equate to the number of days
		// in the week)
		val count = dayOfWeekView.childCount

		// Iterate over each child
		for (i in 0 until count)
		{
			// Get the button
			val button = dayOfWeekView.getChildAt(i) as NacDayButton

			// Set the day
			button.setText(week[i])

			// Set the listener
			button.onDayChangedListener = this
		}
	}

}